### DISCUSSION TITLE
Webcomponents V3

### DISCUSSION THREAD

**Discussion by MarvinKlein1508** [microsoft/fluentui-blazor#2902] [General]
Hey @vnbaaij 

I just thought I might ask you. Is there any roadmap or information about the current state of the webcomponents v3?

I can only find this dev site:
https://web-components.fluentui.dev

And this branch which hasn't been updated in the past 6 months.
https://github.com/microsoft/fluentui/tree/web-components-v3

Where can we get some information about it?

**Comment by dvoituron** [comment]
About the new FluentUI Web Component, you have:
- The current roadmap: https://github.com/orgs/microsoft/projects/652/views/2
- The source code: https://github.com/microsoft/fluentui/tree/master/packages/web-components/src
- Our FluentUI-Blazor branch: https://github.com/microsoft/fluentui-blazor/tree/dev-v5

But no release date has yet been published.

**Comment by MarvinKlein1508** [comment]
Hi @dvoituron 

thanks this helps a lot! :)

**Comment by amather** [comment]
Hey @dvoituron @vnbaaij ,

Just found the thread and wanted ask follow up questions, mainly what you expect from v3 and and what we end-users can expect from the future update?

The way I see it, until now, web-components wrapped fast-foundation, a project that was effectively discontinued and is no longer supported. The new web-components v3 will therefore be a full rewrite implying that incorporating v3 into fluentui-blazor will be a major task for you and a major update for us end users.

Is my view correct? If so, I'd be even more interested in your thoughts... e.g. is it even feasable to plan for a migration guide if changes are that extensive? Or do you expect API compatibility, so impact is managable? 

Thanks!

**Reply by vnbaaij** [reply]
> The way I see it, until now, web-components wrapped fast-foundation, a project that was effectively discontinued and is no longer supported. The new web-components v3 will therefore be a full rewrite implying that incorporating v3 into fluentui-blazor will be a major task for you and a major update for us end users.

Yes, the FAST team deprecated fast-foundation and they are sort of rewriting the web components for v3 (which still uses FAST technology stack internally!). That on it self does not make incorporating v3 in Fluent UI Blazor a major task. What does make it a major task is that a lot of attribute and property names are changing. This is not because of the rewrite but because the Fluent Web Components are getting aligned to the Fluent UI React components in that area.

 
> Is my view correct? If so, I'd be even more interested in your thoughts... e.g. is it even feasable to plan for a migration guide if changes are that extensive? Or do you expect API compatibility, so impact is managable?

For you as end users there will be breaking changes when moving from v4 to v5. It will most definitely not be a matter of installing the new package and be done. Where possible, we will deliver guidance, docs and tools/utilities but it is not our main goal to make this a one-step-super-easy upgrade. For a look at the current status and how we do migration notes, see https://fluentui-blazor-v5.azurewebsites.net/MigrationV5

In fact, I don't think it will be a 'real' upgrade at all in the first period of v5 availability. The Web Components v3 release will **not** include all v2 Web Components in its initial release. The set will grow over time but some components might not return at all. It will strongly depend on what components you are using in your project if you can migrate directly, later or, in a worst case scenario, not at all,

ps. Due to the technical issues it is not/will not be possible to use V2 & V3 Web Components side-by-side

**Reply by MarvinKlein1508** [reply]
> The Web Components v3 release will not include all v2 Web Components in its initial release

@vnbaaij do you already know some confirmed components whhich won't be there from the beginning?

> but some components might not return at all

same question as before

**Reply by vnbaaij** [reply]
No, we don't have an answer to either of those questions yet.

Keep an eye on https://web-components.fluentui.dev/?path=/docs/components-label--docs on what will be in v3 and that associated roadmap on what will come later.

Basically v3+roadmap will give you what will not be ported from v2

**Comment by lksr-demant** [comment]
Hi All,
I have a follow-up question in this matter. My company starts new project based on Blazor. We plan to leverage the FluentUI-Blazor library to create a robust UI. In our project, the Design System will be a crucial aspect, as we need to support multiple themes (brands) that can change dynamically at runtime. Therefore, we require a clear and comprehensive solution that allows customization using DesignTokens.

Currently, the 4.x version of your controls is based on web-components v2.6.1, which leverages the DesignToken object from fast-foundation (https://www.fluentui-blazor.net/DesignTokens). In web-components v3, this approach is completely revamped, which I assume will also change in Blazor FluentUI v5. Am I correct?

Overall, I see a risk in starting the project with an older version of the library that will undergo significant changes soon. Do you have any timelines for when we can expect the new version of your product? Would it be around Q2 of 2025 or rather 2026?
Thank you in advance,
Leszek

**Reply by vnbaaij** [reply]
> Currently, the 4.x version of your controls is based on web-components v2.6.1, which leverages the DesignToken object from fast-foundation (https://www.fluentui-blazor.net/DesignTokens). In web-components v3, this approach is completely revamped, which I assume will also change in Blazor FluentUI v5. Am I correct?


Yes, you are correct on that. 

> Overall, I see a risk in starting the project with an older version of the library that will undergo significant changes soon. Do you have any timelines for when we can expect the new version of your product? Would it be around Q2 of 2025 or rather 2026? Thank you in advance,

The v4 version supports both .NET 8 and 9 and will be supported till those SDKs are supported. We don't have a set timeline for v5. The Web Components v3 are at the moment still in beta and will go through some RCs before a final version is there. I don't have a timeline for that either. What I do know it that the initial set of components in v3 will **not** be the same as what we currently have. This means that v5 will not be a drop-in replacement from a functional perspective. It will **definitely** not be a drop-in from a code perspective and there will be quite a few breaking changes (because of changes in v3 web components which are being aligned to the Fluent React components for parameter names, enums, etc).

So for the foreseeable future I see no risk in starting with v4. It will be supported at least until November 2026 (and it's not like it will stop working then). Once there is a stable v5 release you can investigate if it is worthwhile to go to that version and what is needed to do that.
We have 'customers' like the .NET Aspire Dashboard that need to be kept working as well with what is available now. It can not just be abandoned anymore just like that.

**Comment by MarvinKlein1508** [comment]
@vnbaaij @dvoituron is there any URL for the web components preview? It looks like my previously mentioned one won't get updated anymore. Also how can I tell which beta version of the webcomponents is being used by the v5 branch?

**Reply by vnbaaij** [reply]
The previous URL is still correct. But they are having an issue with their workflows on updating the site.

You can see which version of the components script we use in https://github.com/microsoft/fluentui-blazor/blob/bacb5dbc5d3504f2f00616a6ec6deb7faafb059a/src/Core.Scripts/package.json (currently beta.99, we need to update that to beta.123)
